lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -855,6 +855,92 @@ static void optimizeBatchNorm(Function *F) {
+                                       ConcatNode *origConcatN) {
+  // For the Concat to work, the inputs must only have one
+  // different dim, like <2,1,3>, <2,2,3> can be concatenated
+  // while <2,1,3>, <1,2,3> cannot. So for this function, it tries
+  // to check the inputs fall into the following two cases:
+  //  1). The inputs dims vectors are all the same. e.g. <2,2,3> ... <2,2,3>.
+  //      For this case, the returned dim is the chosen index that can maintain
+  //      the data order as by the original ConcatNode (origConcatN). A example
+  //      is like, the origConcatN concatenates two <4,3> on the 2nd dim, then
+  //      the new Concat can only concatenate on the 1st/2nd dim, as we want the
+  //      concat "granularity" bigger than 3 (which is of the origConcatN).
+  //  2). Only one dim is different for all the dims vectors. e.g. <2,1,3>,
+  //      <2,2,3>, <2,2,3>, <2,3,3>. For this case, the returned dim is this dim  
+  //      i where dims1[i] != dims2[i].
+  auto firstDims = inputs.front().dims();
+  size_t dimSize = firstDims.size();
+  auto origCNInputShape = origConcatN->getInputs().front().dims();
+  int minimalNumOfUnitToLegallyConcat = 1;
+  // Find the minimal number of data units so that we can legally Concat them.
+  for (size_t i = origConcatN->getDim(); i < origCNInputShape.size(); ++i) {
+    minimalNumOfUnitToLegallyConcat *= origCNInputShape[i];
+  }
+
+  ssize_t index = -1;
+  ssize_t preIndex = -1;
+  for (auto &in : inputs) {
+    auto curDims = in.dims();
+    // Find the previous non-negative index value.
+    if (index != -1) {
+      if (preIndex != -1 && preIndex != index) {
+        // We cannot have two different non-negative index to be the
+        // dim for Concat.
+        return -1;
+      }
+      preIndex = index;
+    }
+    index = -1;
+    for (size_t i = 0; i < dimSize; ++i) {
+      // Find the index i where firstDims[i] != curDims[i]
+      if (firstDims[i] != curDims[i]) {
+        if (index != -1) {
+          // If curDims and firstDims have more than one different element ,
+          // then we can not perform Concat on top of these two nodes.
+          return -1;
+        }
+        index = i;
+      }
+    }
+  }
+
+  // Determine the index that can be the dim for Concat.
+  index = index == -1 ? preIndex : index;
+  // It's still possible that index can be -1. it means all inputs have
+  // same shape.
+  if (index == -1) {
+    // In this case, we find the index that can meet the requirement of
+    // minimal number of data units to legally concat.
+    int cumulativeNumUnits = 1;
+    for (ssize_t i = dimSize - 1; i >= 0; i--) {
+      cumulativeNumUnits *= firstDims[i];
+      // Go over the dimensions from right to left, find the first index that
+      // can make cumulativeNumUnits divided by minimalNumOfUnitToLegallyConcat.
+      if (cumulativeNumUnits % minimalNumOfUnitToLegallyConcat == 0) {
```

Comment:
Question: Is it really OK if cumulativeNumUnits are bigger than minimalNumOfUnitToLegallyConcat? Will such a concat really produce the same results or may be it will concat a bit more than expected? 

Essentially, it comes back to your comment `as we want the concat "granularity" bigger than the one of the origConcatN`. I'm trying to understand why it is always safe if it is bigger.

Reply:
I think you are right here. Better to be safe. Changed to check if cumulativeNumUnits == minimalNumOfUnitToLegallyConcat.